runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+		if err != nil {
+			d.logger.Warn("error in closing old read handle", slog.String("error", err.Error()))
+		}
+	}
+
+	// delete data for tables/versions that have been removed from write
+	entries, err = os.ReadDir(d.readPath)
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		readVersion, ok, _ := d.readTableVersion(entry.Name())
+		if !ok {
+			// invalid table
+			_ = os.RemoveAll(filepath.Join(d.readPath, entry.Name()))
+			continue
+		}
+
+		writeVersion, ok := tableVersion[entry.Name()]
+		if !ok {
+			// table not in write
+			d.logger.Debug("Sync: removing table", slog.String("table", entry.Name()))
+			err = os.RemoveAll(filepath.Join(d.readPath, entry.Name()))
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		if readVersion == writeVersion {
+			continue
+		}
+
+		d.logger.Debug("Sync: removing old version", slog.String("table", entry.Name()), slog.String("version", readVersion))
+		err = os.RemoveAll(filepath.Join(d.readPath, entry.Name(), readVersion))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *db) Size() int64 {
+	var paths []string
+	entries, err := os.ReadDir(d.readPath)
+	if err != nil { // ignore error
+		return 0
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		// this is to avoid counting temp tables during source ingestion
+		// in certain cases we only want to compute the size of the serving db files
+		// TODO :: remove this when removing staged table concepts
+		if strings.HasPrefix(entry.Name(), "__rill_tmp_") {
+			continue
+		}
+		path := filepath.Join(d.readPath, entry.Name())
+		version, exist, _ := d.readTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		paths = append(paths, filepath.Join(path, fmt.Sprintf("%s.db", version)))
+	}
+	return fileSize(paths)
+}
+
+// acquireWriteConn syncs the write database, initializes the write handle and returns a write connection.
+// The release function should be called to release the connection.
+// It should be called with the writeMu locked.
+func (d *db) acquireWriteConn(ctx context.Context) (*sqlx.Conn, func() error, error) {
+	err := d.syncWrite(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	db, err := d.openDBAndAttach(ctx, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	conn, err := db.Connx(ctx)
+	if err != nil {
+		_ = db.Close()
+		return nil, nil, err
+	}
+	return conn, func() error {
+		_ = conn.Close()
+		err = db.Close()
+		return err
+	}, nil
+}
+
+func (d *db) openDBAndAttach(ctx context.Context, read bool) (*sqlx.DB, error) {
+	// open the db
+	var (
+		dsn      *url.URL
+		err      error
+		settings map[string]string
+		path     string
+	)
+	if read {
+		dsn, err = url.Parse("") // in-memory
+		settings = d.opts.ReadSettings
+		path = d.readPath
+	} else {
+		path = d.writePath
+		dsn, err = url.Parse(filepath.Join(path, "stage.db"))
+		settings = d.opts.WriteSettings
+	}
```

Comment:
Similar comment as above – did you consider connecting to the specific `version.db` that we're about to create? That would avoid running DML statements on an attached database (which might be fragile based on some of the errors we have seen before).

Reply:
I did not consider it. 
Any specific bugs you have in mind. We have seen some bugs which were due to the overall nature of running duckDB in this fashion but nothing explicitly due to the DML query being on attached db instead of on main db. 
I mean we can obviously change this with few lines of code but thinking if it is necessary.

Reply:
Don't remember specifics, but IIRC when mutating attached databases, it goes into some multi-WAL transaction code paths, doesn't it?

Apart from that, I guess connecting directly to the file we're creating/editing might be more resilient for a) error cases where the previous `stage.db` handle is not completely closed or has an orphan `.wal` file or something like that, b) enabling support for concurrent writes

Reply:
Sure. Though there are some edge cases to handle as I discovered in testing.
1. When replacing table we need to make sure we attach existing table (so that views if any are properly created) and then drop the view (`SELECT * FROM table.table`) so that a table can be created and does not conflict with existing view.
2. In mutate table we need to ignore this table while attaching because the table is already present in db file.